_posts/2017-10-05-progress-report-q3-17.md

```diff
@@ -0,0 +1,80 @@
+parts of the library to FreeBSD, mostly led by [@gaweinbergi](https://github.com/gaweinbergi).
+Our code is naturally very linux centric due to being developed
+alongside the kernel support for NVDIMMs, and there are quite a few
+hurdles to overcome, but everything seems to be going along quite nicely.
+- Patch that ports the low-level library (libpmem) to ARM has been
+submitted to our repository [@vvenkates27](https://github.com/vvenkates27).
+This is also a place where there's a lot of platform specific code due to
+the novelty of the topic and the fact that the library was implemented
+together with the programming model for x86.
+
+### Allocation classes merged
+
+I've talked about this previously, so I won't repeat myself here, but the
+allocation classes were finally merged last quarter.
+
+For anyone to whom this is a topic of interest, I suggest looking at the
+[slab
+allocator example](https://github.com/pmem/nvml/tree/master/src/examples/libpmemobj/slab_allocator)
+that shows the exact use case this feature is intended for.
+
+### Optimized startup time
+
+Here's a short, but hopefully interesting, bit of news: the happy-case
+performance of `pmemobj_open()` has increased by two orders of magnitude due to
+strategic usage of lazy initialization.
+
+### Bad blocks & high-availability features
+
+Just as in traditional disks, NVDIMMs can have bad physical blocks on the medium.
+The specification on how those blocks are exposed has been recently published
+with [ACPI 6.2](http://www.uefi.org/sites/default/files/resources/ACPI_6_2.pdf).
+
+From linux kernel file system perspective, the NVDIMM badblocks are exposed
+like in traditional storage, and can be cleared through normal fs interfaces.
+
+The story changes a bit when the memory is mapped in the address space of the
+application - unsurprisingly, the semantics are similar to those of bad volatile
+memory - a `SIGBUS` with appropriate flags is generated if a poisoned (bad) page
+is touched by the program.
+
+Regular software doesn't really have to worry about poisoned pages because memory
+is volatile and a `SIGBUS` normally terminates the application, and when started
+again, everything will be back in order.
+
+Our initial approach focuses on letting the application terminate when a `SIGBUS`
+is encountered, and then provide a set of off-line tools that allow seamless
+recovery of data from replicas.
+
+The work is still in progress and the relevant PR is [here](https://github.com/pmem/nvml/pull/2246).
+
+### Conclusion
+
+Vacation season might appear to have taken the toll this time around, but that's
+mostly just that - an appearance. 
+
+The vast majority of changes this quarter were fixes and tiny improvements.
+
+For more exciting developments, look at [pmemkv](https://github.com/pmem/pmemkv), [pynvm](https://github.com/pmem/pynvm) or [pmse](https://github.com/pmem/pmse) where a lot of activity has been
+concentrated around.
+
+Until the next time.
```

Comment:
I think it's _Until next time_.

Reply:
Hm, is using this idiom on its own ever grammatically correct? I got rid of it entirely.